[KB] Match comments with generic mode (#1312)
semgrep/semgrep-docs: docs/kb/rules/match-commments.md

---
description: Semgrep's generic pattern matching mode can match comments in code files.
tags:
  - Rules
  - Semgrep Code
---

# Match comments with Semgrep

> When Semgrep rules target a language or languages, they do not match comments in the targeted code files. Comments are not part of the semantic and syntactic structure of the document, so they are ignored.

Reviewer: ```suggestion
When Semgrep rules target languages, they do not match comments in the targeted code files. Comments are not part of the semantic and syntactic structure of the document, so they are ignored.
```

Author: good call that this is awkward - I think I might go with "target specific languages" to make clear the opposition to `generic`

Author: maybe I can just suggest things to myself, instead of doing the proper git thing... 😂

> When Semgrep rules target a language or languages, they do not match comments in the targeted code files. Comments are not part of the semantic and syntactic structure of the document, so they are ignored.

Author: ```suggestion
When Semgrep rules target specific languages, they do not match comments in the targeted code files. Comments are not part of the semantic and syntactic structure of the document, so they are ignored.
```

However, it's sometimes useful to match comments. For example, comments can control the behavior of other linters, such as type checkers. You might also have certain formatting standards for comments, such as requiring that a `TODO` comment contains a ticket capturing the required work.

To match comments with Semgrep, use the `generic` language target to invoke [generic pattern matching](/docs/writing-rules/generic-pattern-matching/).

## Example rule

> Suppose that your organization requires all `TODO` comments to have an associated Jira ticket. This rule finds TODO lines with no `atlassian.net` content. It would identify any lines not containing a Jira Cloud ticket link.

Reviewer: ```suggestion
Suppose that your organization requires all `TODO` comments to have an associated Jira ticket. This rule finds TODO lines with no `atlassian.net` content and identifies any lines not containing a Jira Cloud ticket link.
```

```yaml
rules:
  - id: no-todo-without-jira
    patterns:
      - pattern: TODO $...ACTION
      - pattern-not: TODO ... atlassian.net ...
    options:
      generic_ellipsis_max_span: 0
    message: The TODO comment "$...ACTION" does not contain a Jira ticket to resolve the issue
    languages:
      - generic
    severity: INFO
    metadata:
      category: best-practice
```

:::note
Try this pattern in the [Semgrep Playground](https://semgrep.dev/playground/s/lBDRL). 
:::

> This rule also includes the `generic_ellipsis_max_span` option, which [limits the ellipsis to matching on the same line](/docs/writing-rules/generic-pattern-matching/#handling-line-based-input), and prevents it from over-matching in this generic context.

Reviewer: ```suggestion
This rule also includes the `generic_ellipsis_max_span` option, which [limits the ellipsis to matching on the same line](/docs/writing-rules/generic-pattern-matching/#handling-line-based-input) and prevents it from over-matching in this generic context.
```
